File: 剑指offer/python/28.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def is_cur(L, R):
            if L == None and R == None: 
                return True
            if L == None or R == None or L.val != R.val: 
                return False
            return is_cur(L.left, R.right) and is_cur(L.right, R.left)

        return is_cur(root.left, root.right) if root else True

File: 剑指offer/cpp/28.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        return is_curr(root->left, root->right);
    }
    bool is_curr(TreeNode* r1, TreeNode* r2) {
        if (!(r1 || r2)) return true; // 两个都是null
        if (!(r1 && r2) || (r1->val != r2->val)) return false;
        return is_curr(r1->left, r2->right) && is_curr(r1->right, r2->left);
    }

};